### ISSUE TITLE
Potential file handle leak since mill 0.5.7

### ISSUE THREAD

**Issue by lefou** [lihaoyi/mill#754]
This first popped up in gitter: https://gitter.im/lihaoyi/mill?at=5e16ff37e52f635378ad933a

The initial stack trace:

```

[1585/1733] framework.rmfApp.allSourceFiles 
1 targets failed
framework.rmfApp.allSourceFiles java.nio.file.FileSystemException: <redacted>/framework/rmf.app/src/main/java: Zu viele offene Dateien
    sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
    sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
    sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
    sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:427)
    java.nio.file.Files.newDirectoryStream(Files.java:457)
    os.walk$stream$$anon$2.generate(ListOps.scala:225)
    geny.Generator$Mapped.generate(Generator.scala:283)
    geny.Generator.foreach(Generator.scala:49)
    geny.Generator.foreach$(Generator.scala:49)
    geny.Generator$Mapped.foreach(Generator.scala:281)
    geny.Generator.toBuffer(Generator.scala:127)
    geny.Generator.toBuffer$(Generator.scala:125)
    geny.Generator$Mapped.toBuffer(Generator.scala:281)
    geny.Generator.toArray(Generator.scala:131)
    geny.Generator.toArray$(Generator.scala:131)
    geny.Generator$Mapped.toArray(Generator.scala:281)
    os.walk$.apply(ListOps.scala:95)
    mill.scalalib.JavaModule.$anonfun$allSourceFiles$4(JavaModule.scala:193)
    scala.collection.TraversableLike$WithFilter.$anonfun$flatMap$2(TraversableLike.scala:773)
    scala.collection.Iterator.foreach(Iterator.scala:941)
    scala.collection.Iterator.foreach$(Iterator.scala:941)
    scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
    scala.collection.IterableLike.foreach(IterableLike.scala:74)
    scala.collection.IterableLike.foreach$(IterableLike.scala:73)
    scala.collection.AbstractIterable.foreach(Iterable.scala:56)
    scala.collection.TraversableLike$WithFilter.flatMap(TraversableLike.scala:772)
    mill.scalalib.JavaModule.$anonfun$allSourceFiles$2(JavaModule.scala:191)
    mill.define.Task$MappedDest.evaluate(Task.scala:365)
```

A short run of `lsof -p` (on Linux) on the mill server process after mill client finished suggests, that mill keeps all `meta.json` files open. 

Run in mill repo with mill 0.5.7:
```
23:42:01 ⌁0% θ74° lefou:~/work/opensource/mill master(1)+* ± mill assembly; lsof -p 7552 | wc
[2/301] publishVersion 
[274/301] dev.sources 
   3622   32598  523167
23:42:03 ⌁0% θ74° lefou:~/work/opensource/mill master(1)+* ± mill assembly; lsof -p 7552 | wc
[2/301] publishVersion 
[274/301] dev.sources 
   3911   35199  564671
23:42:05 ⌁0% θ74° lefou:~/work/opensource/mill master(1)+* ± mill assembly; lsof -p 7552 | wc
[2/301] publishVersion 
[281/301] dev.unmanagedClasspath 
Exception in thread "MillServerActionRunner" 
   4178   37602  603159
23:42:07 ⌁0% θ74° lefou:~/work/opensource/mill master(1)+* ± mill assembly; lsof -p 7552 | wc
java.io.FileNotFoundException: out/mill-worker-to2CVwVb3b2Ba/50wr+rtQIqC2I=-1/run (Zu viele offene Dateien)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at java.io.FileInputStream.<init>(FileInputStream.java:93)
        at mill.main.Server.handleRun(MillServerMain.scala:124)
        at mill.main.Server.$anonfun$run$2(MillServerMain.scala:105)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.lockBlock(MillServerMain.scala:198)
        at mill.main.Server.$anonfun$run$1(MillServerMain.scala:84)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.tryLockBlock(MillServerMain.scala:205)
        at mill.main.Server.run(MillServerMain.scala:81)
        at mill.main.MillServerMain$.main(MillServerMain.scala:45)
        at mill.main.MillServerMain.main(MillServerMain.scala)
Exception in thread "main" java.io.IOException: com.sun.jna.LastErrorException: [24] Zu viele offene Dateien
        at org.scalasbt.ipcsocket.UnixDomainServerSocket.<init>(UnixDomainServerSocket.java:112)
        at org.scalasbt.ipcsocket.UnixDomainServerSocket.<init>(UnixDomainServerSocket.java:93)
        at mill.main.Server.$anonfun$run$2(MillServerMain.scala:91)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.lockBlock(MillServerMain.scala:198)
        at mill.main.Server.$anonfun$run$1(MillServerMain.scala:84)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.tryLockBlock(MillServerMain.scala:205)
        at mill.main.Server.run(MillServerMain.scala:81)
        at mill.main.MillServerMain$.main(MillServerMain.scala:45)
        at mill.main.MillServerMain.main(MillServerMain.scala)
Caused by: com.sun.jna.LastErrorException: [24] Zu viele offene Dateien
        at org.scalasbt.ipcsocket.UnixDomainSocketLibrary.socket(Native Method)
        at org.scalasbt.ipcsocket.UnixDomainServerSocket.<init>(UnixDomainServerSocket.java:103)
        ... 10 more
      0       0       0
```

This does not happen with mill 0.5.3 and 0.5.3-13-4f4fab

**Comment by lefou** [comment]
After bisecting the issue, I found the offending commit: 512f4933cb8b5bf0fea520637d538f869cfa10c3

```
512f4933cb8b5bf0fea520637d538f869cfa10c3 is the first bad commit
commit 512f4933cb8b5bf0fea520637d538f869cfa10c3
Author: Li Haoyi <haoyi.sg@gmail.com>
Date:   Fri Dec 27 12:55:18 2019 +0800

    Bump versions again (#750)
    
    Pull in latest versions of Ammonite, OS-Lib, uPickle, Sourcecode

 build.sc                                     | 10 +++++-----
 main/api/src/mill/api/JsonFormatters.scala   |  4 ++--
 main/core/src/eval/Evaluator.scala           |  4 ++--
 scalalib/src/TestRunner.scala                |  2 +-
 scalalib/src/publish/SonatypeHttpApi.scala   |  8 ++++----
 scalalib/src/publish/SonatypePublisher.scala |  2 +-
 6 files changed, 15 insertions(+), 15 deletions(-)
```

**Comment by lefou** [comment]
@lihaoyi Could you please have a look?

**Comment by lihaoyi** [comment]
yeah i'll take a look, thanks for helping bisect!

**Comment by lihaoyi** [comment]
@lefou do you have a convenient way I can use to reproduce the issue? Like run some command over and over?

**Comment by lefou** [comment]
I just re-run `mill assembly` in the mill repo. Although everthing is up-to-date, the count of open file handles increases by around 300. It will break when the limit is reached, so it depends on your system configuration.

**Comment by lefou** [comment]
It's `~/mill-release assembly` to be correct. `~/mill-release` is the outcome of `sh ci/publish-local.sh` in the mill repo. It's is essential to NOT use `-i` option, to keep a mill server running.

**Comment by lihaoyi** [comment]
got it, thanks!

**Comment by lefou** [comment]
FTR, mill-0.5.8 is also affected.

```
10:15:11 ⌁ θ62° 1& lefou:~/work/opensource/mill master(+2/-2)+* 130 ± for i in {1..20} ; do echo $i; mill --mill-version 0.5.8 assembly ; done
1
Mill version changed (0.5.8-2-1240d1-DIRTYfd2d8e37 -> 0.5.8), re-starting server
Compiling (synthetic)/ammonite/predef/interpBridge.sc
Compiling (synthetic)/ammonite/predef/sourceBridge.sc
Compiling (synthetic)/ammonite/predef/frontEndBridge.sc
Compiling /home/lefou/work/opensource/mill/build.sc
[101/301] main.client.compile 
[info] Compiling 9 Java sources to /home/lefou/work/opensource/mill/out/main/client/compile/dest/classes ...
[info] Done compiling.
[301/301] assembly 
2
[2/301] publishVersion 
[274/301] dev.sources 
3
[2/301] publishVersion 
[274/301] dev.sources 
4
[2/301] publishVersion 
[274/301] dev.sources 
5
[2/301] publishVersion 
[274/301] dev.sources 
6
[2/301] publishVersion 
[274/301] dev.sources 
7
[2/301] publishVersion 
[274/301] dev.sources 
8
[2/301] publishVersion 
[274/301] dev.sources 
9
[2/301] publishVersion 
[274/301] dev.sources 
10
[2/301] publishVersion 
[274/301] dev.sources 
11
[2/301] publishVersion 
[274/301] dev.sources 
12
[2/301] publishVersion 
[274/301] dev.sources 
13
[2/301] publishVersion 
[274/301] dev.sources 
14
[2/301] publishVersion 
[250/301] scalanativelib.sources 
Exception in thread "MillServerActionRunner" java.nio.file.FileSystemException: /home/lefou/work/opensource/mill/out/scalanativelib/sources/log: Zu viele offene Dateien

Exception: java.lang.ThreadDeath thrown from the UncaughtExceptionHandler in thread "MillServerActionRunner"
15
java.io.FileNotFoundException: out/mill-worker-to2CVwVb3b2Ba/50wr+rtQIqC2I=-1/run (Zu viele offene Dateien)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at java.io.FileInputStream.<init>(FileInputStream.java:93)
        at mill.main.Server.handleRun(MillServerMain.scala:124)
        at mill.main.Server.$anonfun$run$2(MillServerMain.scala:105)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.lockBlock(MillServerMain.scala:198)
        at mill.main.Server.$anonfun$run$1(MillServerMain.scala:84)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at mill.main.Server$.tryLockBlock(MillServerMain.scala:205)
        at mill.main.Server.run(MillServerMain.scala:81)
        at mill.main.MillServerMain$.main(MillServerMain.scala:45)
        at mill.main.MillServerMain.main(MillServerMain.scala)
```

**Comment by lefou** [comment]
This bug is really annoying. on a larger project it happens on almost any second build. We are actually forced to stay on 0.5.3-x because of this.

I added a test script under `ci/file-leak-test.sh` for anyone who want's to reproduce.

**Comment by ajrnz** [comment]
I've just had a quick look at this. It appears to be a problem somewhere under `upickle.default.read`.

Here is a minimized test case:

```scala
object MillTest extends App {
  //val baseDir = os.pwd
  val baseDir = os.home / "dev" / "ext" / "mill"
  val x =
    os.walk(baseDir)
      .filter(_.last == "meta.json")
      .map{ p =>
      println(s"Loading: $p")
      upickle.default.read[ujson.Value](p.toIO)
    }

  println("waiting")
  Thread.sleep(1000*1000)
}
```
Using:

```
    ivy"com.lihaoyi::upickle:0.9.5",
    ivy"com.lihaoyi::os-lib:0.6.2"
```

If you find the java PID and then do a `lsof -p <pid>` you see all the files are open.

Appears to have been introduced between upickle 0.9.0 and 0.9.5.

Don't have more time to dig at the moment but hope this helps...

**Comment by lefou** [comment]
Thanks @ajrnz! I opened an issue upstream: https://github.com/lihaoyi/upickle/issues/303

**Comment by nafg** [comment]
I guess now we just need a new release of upickle and update Mill to it?

**Comment by lefou** [comment]
@nafg, thank you! I will update the dependency once new upickle is released. Which version will it be? Last time I checked upickle release versions and public git tags were out of sync. ()

**Comment by nafg** [comment]
@lefou https://github.com/lihaoyi/upickle/pull/304#issuecomment-590009917 says it was published as upickle 1.0.0

**Comment by lefou** [comment]
Thanks to all who helped!

**Comment by sequencer** [comment]
Did we fixed this bug in 0.6.1, I possible encounter this bug in my `0.6.1`, when I try to write a plugin.
